packages/core/src/stream-loading/state-manipulator.ts

```diff
@@ -1,27 +1,266 @@
+    return log.findIndex((logEntry) => logEntry.type == CommitType.ANCHOR)
   }
 
+  /**
+   * Given two different Stream logs representing two different conflicting histories of the same
+   * Stream, pick which history to accept, in accordance with our conflict resolution strategy.
+   * The inputted logs should contain only the new commits past the divergence point between the
+   * two histories - there should be no commits in common between the two input logs.
+   * @param log1
+   * @param log2
+   * @returns the log that is selected by the conflict resolution rules.
+   */
+  _pickLogToAccept(log1: Array<LogEntry>, log2: Array<LogEntry>): Array<LogEntry> {
+    const firstAnchorIndexForLog1 = this._findAnchorIndex(log1)
+    const firstAnchorIndexForLog2 = this._findAnchorIndex(log2)
+    const isLog1Anchored = firstAnchorIndexForLog1 >= 0
+    const isLog2Anchored = firstAnchorIndexForLog2 >= 0
+
+    // When one of the logs is anchored but not the other, take the one that is anchored
+    if (isLog1Anchored != isLog2Anchored) {
+      return isLog1Anchored ? log1 : log2
+    }
+
+    if (isLog1Anchored && isLog2Anchored) {
+      // When both logs are anchored, take the one anchored first.
+      const anchorTimestamp1 = log1[firstAnchorIndexForLog1].timestamp
+      const anchorTimestamp2 = log2[firstAnchorIndexForLog2].timestamp
+      if (anchorTimestamp1 < anchorTimestamp2) {
+        return log1
+      } else if (anchorTimestamp2 < anchorTimestamp1) {
+        return log2
+      }
+    }
+
+    // When both logs are anchored in the same block (or neither log is anchored), compare log
+    // lengths until that anchor (or the end of the log if not anchored) and choose the one with
+    // longer length.
+    // TODO(CDB-2746) - it's kind of dumb that we only consider the log up until the first anchor.
+    // This is basically a holdover from the way conflict resolution was originally implemented, but
+    // changing it now would be a breaking change.
+    const relevantLength1 = isLog1Anchored ? firstAnchorIndexForLog1 + 1 : log1.length
+    const relevantLength2 = isLog2Anchored ? firstAnchorIndexForLog2 + 1 : log2.length
+
+    if (relevantLength1 > relevantLength2) {
+      return log1
+    } else if (relevantLength1 < relevantLength2) {
+      return log2
+    }
+
+    // If we got this far, that means that we don't have sufficient information to make a good
+    // decision about which log to choose.  The most common way this can happen is that neither log
+    // is anchored, although it can also happen if both are anchored but in the same blockNumber or
+    // blockTimestamp. At this point, the decision of which log to take is arbitrary, but we want it
+    // to still be deterministic. Therefore, we take the log whose last entry has the lowest CID.
+    return log1[log1.length - 1].cid.bytes < log2[log2.length - 1].cid.bytes ? log1 : log2
+  }
+
+  async _applyLogToState_noCacaoVerification(
```

Comment:
Not a big deal but I don't really like mixing casing conventions like this, couldn't this method be named `_applyLogToStateWithoutCacaoVerification` for example please?